Source/WebCore/inspector/InspectorStyleSheet.cpp

```diff
@@ -1436,23 +1471,90 @@ Ref<Protocol::CSS::CSSStyle> InspectorStyleSheet::buildObjectForStyle(CSSStyleDe
     return result;
 }
 
-ExceptionOr<void> InspectorStyleSheet::setStyleText(const InspectorCSSId& id, const String& text, String* oldText)
+static inline bool isNotSpaceOrTab(UChar character)
 {
-    auto inspectorStyle = inspectorStyleForId(id);
-    if (!inspectorStyle)
+    return character != ' ' && character != '\t';
+}
+
+ExceptionOr<void> InspectorStyleSheet::setRuleStyleText(const InspectorCSSId& id, const String& newText, String* oldText, bool isUndo)
+{
+    RefPtr<CSSRule> cssRule = ruleForId(id);
+    if (!cssRule)
         return Exception { NotFoundError };
 
+    RefPtr<CSSRuleSourceData> sourceData = ruleSourceDataFor(cssRule.get());
+    if (!sourceData)
+        return Exception { NotFoundError };
+
+    RefPtr<CSSRuleSourceData> logicalContainingRuleSourceData = sourceData->isImplicitlyNested ? ruleSourceDataFor(cssRule->parentRule()) : sourceData;
+    if (!logicalContainingRuleSourceData)
+        return Exception { NotFoundError };
+
+    unsigned bodyStart = logicalContainingRuleSourceData->ruleBodyRange.start;
+    unsigned bodyEnd = logicalContainingRuleSourceData->ruleBodyRange.end;
+    ASSERT(bodyStart <= bodyEnd);
+
+    String styleSheetText = m_parsedStyleSheet->text();
+    RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(bodyEnd <= styleSheetText.length());
+
     if (oldText) {
-        auto result = inspectorStyle->text();
-        if (result.hasException())
-            return result.releaseException();
-        *oldText = result.releaseReturnValue();
+        // In order to perform a faithful undo, text has to be restored to its non-canonicalized form.
+        *oldText = styleSheetText.substring(bodyStart, bodyEnd - bodyStart);
     }
 
-    auto result = inspectorStyle->setText(text);
-    if (!result.hasException())
-        fireStyleSheetChanged();
-    return result;
+    auto setPatchedStyleSheetText = [&](String& patchedStyleSheetText) {
+        setText(patchedStyleSheetText);
+        reparseStyleSheet(patchedStyleSheetText);
+    };
+
+    if (isUndo) {
+        // Undo operations will be performed with complete style text, including nested rules.
+        auto patchedStyleSheetText = makeStringByReplacing(styleSheetText, bodyStart, bodyEnd - bodyStart , newText);
+        setPatchedStyleSheetText(patchedStyleSheetText);
+        return { };
+    }
+
+    auto indentation = emptyString();
+    auto startOfSecondLine = newText.find('\n');
+    if (startOfSecondLine != notFound) {
+        ++startOfSecondLine;
+        auto endOfSecondLineWhitespace = newText.find(isNotSpaceOrTab, startOfSecondLine);
+        if (endOfSecondLineWhitespace != notFound)
+            indentation = newText.substring(startOfSecondLine, endOfSecondLineWhitespace - startOfSecondLine);
+    }
```

Comment:
I'm hopefully someone has a better way to accomplish this and the indentation groking below 😞

Reply:
i wonder if we should maybe change `Inspector::ContentSearchUtilities::countRegularExpressionMatches` to instead be `getRegularExpressionMatches` and find the most common (or average) line whitespace prefix

Reply:
Because we get the properties pre-formatted from the frontend, the average of everything but the first and last line should always be the same as just getting the indentation of a single line. The last line is then some fraction of that average, and the first line should always have no space (since the "text" is everything between the `{}`.